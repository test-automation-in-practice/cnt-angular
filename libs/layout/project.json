{
  "name": "layout",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "library",
  "sourceRoot": "libs/layout/src",
  "prefix": "cntws",
  "targets": {
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/layout"],
      "options": {
        "jestConfig": "libs/layout/jest.config.ts"
      }
    },
    "lint": {
      "executor": "@nx/eslint:lint"
    }
  },
  "tags": []
}
