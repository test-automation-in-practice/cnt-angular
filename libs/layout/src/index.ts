export * from './lib/layout.module';
export * from './lib/model/navigation.model';
