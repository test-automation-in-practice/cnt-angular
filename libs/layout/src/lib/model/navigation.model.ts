import {Type} from "@angular/core";

export interface NavbarEntry {
  displayName: string;
  buildRouterLink: () => string;
}

export class DefinedRoute implements NavbarEntry {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(public path: string, public component: Type<any>, public displayName: string) {}

  buildRouterLink() {
    return '/' + this.path.replace('**', '');
  }
}
