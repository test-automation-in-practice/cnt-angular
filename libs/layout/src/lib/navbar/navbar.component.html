<mat-toolbar color="primary" class="w-screen">
  <a *ngFor="let route of navbarEntries" [routerLink]="route.buildRouterLink()" mat-button>
    {{ route.displayName }}
  </a>
</mat-toolbar>
