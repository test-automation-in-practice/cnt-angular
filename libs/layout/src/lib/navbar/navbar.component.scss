:host {
  @apply sticky top-0 left-0 z-10 w-full mb-8;
}
