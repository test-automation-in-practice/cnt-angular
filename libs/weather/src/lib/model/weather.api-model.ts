export interface WeatherApiModel {
  id: number;
  temp: number;
  name: string;
}

export interface LocationApiModel {
  name: string;
}
