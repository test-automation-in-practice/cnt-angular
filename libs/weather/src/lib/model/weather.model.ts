export interface WeatherLocation {
  id: number;
  temp: number;
  location: string;
}
