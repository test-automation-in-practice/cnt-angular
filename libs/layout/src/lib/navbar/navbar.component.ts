import { Component, Input } from '@angular/core';
import { NavbarEntry } from '../model/navigation.model';

@Component({
  selector: 'cntws-navbar',
  templateUrl: './navbar.component.html',
  styleUrls: ['./navbar.component.scss'],
})
export class NavbarComponent {
  @Input() navbarEntries: NavbarEntry[] = [];
}
