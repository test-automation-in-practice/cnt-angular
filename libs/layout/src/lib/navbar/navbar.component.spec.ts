import { MockBuilder, MockRender } from 'ng-mocks';
import { NavbarComponent } from './navbar.component';
import { LayoutModule } from '../layout.module';
import { NavbarEntry } from '../model/navigation.model';
import { RouterTestingModule } from '@angular/router/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CommonModule, Location } from '@angular/common';
import { MatToolbarModule } from '@angular/material/toolbar';
import { MatButtonModule } from '@angular/material/button';
import { HarnessLoader } from '@angular/cdk/testing';
import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
import { Component } from '@angular/core';
import { MatButtonHarness } from '@angular/material/button/testing';
import { RouterModule } from '@angular/router';

@Component({
  selector: 'cntws-mock',
  template: '<span>mock</span>',
})
class MockComponent {}

describe('Navbar: A user', () => {
  describe('viewing the navbar', () => {
    beforeEach(() => MockBuilder(NavbarComponent, LayoutModule).replace(RouterModule, RouterTestingModule));

    it('should have a consistent layout', () => {
      const fixture = MockRender(NavbarComponent);
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('clicking on a navbar button', () => {
    let fixture: ComponentFixture<NavbarComponent>;
    let component: NavbarComponent;
    let loader: HarnessLoader;
    let location: Location;

    const navbarEntries: NavbarEntry[] = [
      { displayName: 'First', buildRouterLink: () => 'first' },
      { displayName: 'Second', buildRouterLink: () => 'second' },
    ];
    const routes = navbarEntries.map((route) => ({ path: route.buildRouterLink(), component: MockComponent }));

    beforeEach(async () =>
      TestBed.configureTestingModule({
        declarations: [NavbarComponent, MockComponent],
        imports: [CommonModule, RouterTestingModule.withRoutes(routes), MatToolbarModule, MatButtonModule],
      }).compileComponents()
    );

    beforeEach(() => {
      fixture = TestBed.createComponent(NavbarComponent);
      component = fixture.componentInstance;
      component.navbarEntries = navbarEntries;
      loader = TestbedHarnessEnvironment.loader(fixture);
      location = TestBed.inject(Location);
    });

    it('should be navigated to the according route', async () => {
      const navbarEntry = navbarEntries[0];
      const button = await loader.getHarness(MatButtonHarness.with({ text: navbarEntry.displayName }));
      await button.click();
      expect(location.path()).toEqual(`/${navbarEntry.buildRouterLink()}`);
    });
  });
});
