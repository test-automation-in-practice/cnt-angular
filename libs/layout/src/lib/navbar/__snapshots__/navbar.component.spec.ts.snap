// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar: A user viewing the navbar should have a consistent layout 1`] = `
<mock-render>
  <cntws-navbar>
    <mat-toolbar
      class="w-screen"
      color="primary"
    >
    </mat-toolbar>
  </cntws-navbar>
</mock-render>
`;
